app/expirynotify.go

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"github.com/mattermost/mattermost-server/v5/mlog"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/utils"
+)
+
+const (
+	OneHourMillis = 60 * 60 * 1000
+)
+
+// NotifySessionsExpired is called periodically from the job server to notify any mobile sessions that have expired.
+func (a *App) NotifySessionsExpired() *model.AppError {
+	if *a.Config().EmailSettings.SendPushNotifications {
+		pushServer := *a.Config().EmailSettings.PushNotificationServer
+		if license := a.License(); pushServer == model.MHPNS && (license == nil || !*license.Features.MHPNS) {
+			mlog.Warn("Push notifications are disabled. Go to System Console > Notifications > Mobile Push to enable them.")
+			return nil
+		}
+	}
+
+	// Get all mobile sessions that expired within the last hour.
+	sessions, err := a.srv.Store.Session().GetSessionsExpired(OneHourMillis, true, true)
+	if err != nil {
+		return err
+	}
+
+	msg := &model.PushNotification{
+		Category: model.CATEGORY_CAN_REPLY,
+		Version:  model.PUSH_MESSAGE_V2,
+		Type:     model.PUSH_TYPE_SESSION,
+	}
+
+	for _, session := range sessions {
+		tmpMessage := msg.DeepCopy()
+		tmpMessage.SetDeviceIdAndPlatform(session.DeviceId)
+		tmpMessage.AckId = model.NewId()
+		tmpMessage.Message = a.getSessionExpiredPushMessage(session)
+
+		errPush := a.sendToPushProxy(tmpMessage, session)
```

Comment:
Just a small comment: `sendToPushProxy should not be directly called because it breaks the concurrency limits that we are trying to enforce here https://github.com/mattermost/mattermost-server/pull/14758. I will send a PR myself once this is merged. This is more of a quick FYI.

Reply:
I see.  The higher level APIs assume a Post notification is being sent and weren't suitable for this.  `sendToPushProxy` method should probably get comments explaining the conditions of use.  Thanks.

Reply:
100% agree. It's in a bad shape. The overall plan is to move push notification to it's separate package. Let's hope it happens in my lifetime.